gateway/src/apicast/cli/command/start.lua

```diff
@@ -110,10 +110,15 @@ local function openresty_binary(candidates)
 end
 
 local function build_env(options, config, context)
+    local configuration_cache = options.cache or context.configuration_cache
+    if configuration_cache then -- Avoid calling tostring() on nil, would return "nil".
+        configuration_cache = tostring(configuration_cache)
```

Comment:
Maybe it is not necessary to call `tostring` here because it is done in the `update_env` function?

Reply:
I think it's needed. The result of this method is not only used in `update_env`, it's also used in the call to `exec` at the end of `mt:__call`.

Reply:
https://github.com/3scale/apicast/blob/30984b947c3abaef6335d5dbb591e0136c07f247/gateway/src/apicast/cli/command/start.lua#L174

Reply:
But `nil` keys are being removed from the table, so it won't even get there.

Reply:
`update_env` does not mutate the table that it receives in the params. so if we do not apply `tostring` here, it will never be applied before it reaches that `exec`.

Reply:
exec does not call `tostring` on the value: https://github.com/3scale/lua-resty-execvp/blob/bfa07e3f82fdd8bfb41be5c152438f03ccd60b9f/src/resty/execvp.lua#L48
and `update_env` does: https://github.com/3scale/apicast/blob/30984b947c3abaef6335d5dbb591e0136c07f247/gateway/src/apicast/cli/command/start.lua#L47 

that means the values in this table should be either a `string` or `nil`. 
lua tables don't have `nil` values, so assigning nil will make the key disappear so it is never iterated by `pairs`

Reply:
Not sure I follow. I probably didn't explain this well. Let me try:
1) `build_env` constructs a table `t` and returns it. We're discussing whether we should call `tostring` in some values of `t` here or not. I think we should.
2) The table `t` returned in the previous step is sent to `update_env`. But that function _does not_ mutate `t`.
3) `t` is sent as a param of the call to `exec`. As 2) didn't modify it, it will be sent as we left it in 1). That's why we need to call `tostring()` in `build_env`. Otherwise, `exec` might receive a table that contains things that are not strings.

Reply:
Looks like I was really confused 😄 


We are saying the same. This table now needs to contain `nil` or `string` and that is what you are ensuring. 

I think it would be good if we could clean it up by changing the `exec` to serialize the env: 
https://github.com/3scale/lua-resty-execvp/blob/bfa07e3f82fdd8bfb41be5c152438f03ccd60b9f/src/resty/execvp.lua#L48

Then we don't have to really think about the serialization anymore.

Reply:
Right. I think that fixes the root problem: https://github.com/3scale/lua-resty-execvp/pull/1